### PR TITLE
Enable FusedBatchNormV3 ops to support 5D tensors

### PR DESCRIPTION
**Pull request by kaixih**
This PR enables the FusedBatchNormV3 and FusedBatchNormGradV3 to support 5D tensors.

Some potential benefits from this PR: 
(1) The 3D convolution networks using batch norm can use the FusedBatchNormV3 and benefit from cuDNN on GPU.
(2) This PR doesn't take the reshape-based method (where we convert to 4D tensors using tf.reshape and then call nn.batch_normalization), because the inserted Reshape ops might prevent the grappler layout optimization pass from cancelling off the adjacent NHWC<->NCHW ops around batch normalization.

FYI. @benbarsdell @nluehr

### CODE REVIEW

File: tensorflow/python/keras/layers/normalization_test.py [hunk 1]
```diff
@@ -319,7 +319,7 @@ def test_v2_fused_attribute(self):
     norm = normalization_v2.BatchNormalization(fused=True)
     self.assertEqual(norm.fused, True)
     inp = keras.layers.Input(shape=(4, 4))
-    with self.assertRaisesRegex(ValueError, '4D input tensors'):
+    with self.assertRaisesRegex(ValueError, '4D or 5D input tensors'):
```
Comment (reedwm) [inline tensorflow/python/keras/layers/normalization_test.py:322]: Can you also add a test to this file? I think adding a test case to the `test_basic_batchnorm` is sufficient.

File: tensorflow/python/ops/nn_grad.py [hunk 2]
```diff
@@ -897,21 +897,28 @@ def _BaseFusedBatchNormGrad(op, version, *grad):
     if data_format == b"NCHW":
       x = array_ops.transpose(x, [0, 2, 3, 1])
       grad_y = array_ops.transpose(grad_y, [0, 2, 3, 1])
+    elif data_format == b"NCDHW":
+      x = array_ops.transpose(x, [0, 2, 3, 4, 1])
+      grad_y = array_ops.transpose(grad_y, [0, 2, 3, 4, 1])
+    target_data_format = ("NHWC" if data_format == b"NCHW" or
```
Comment (reedwm) [inline tensorflow/python/ops/nn_grad.py:903]: the data_format check should be `data_format in (b"NCHW", b"NHWC")`

**Review by reedwm — CHANGES_REQUESTED** [review]
@rmlarsen can you review the grappler files? I reviewed all the non-grappler files.

File: tensorflow/python/keras/layers/normalization.py [hunk 3]
```diff
@@ -330,13 +330,13 @@ def build(self, input_shape):
       # output back to its original shape accordingly.
       if self._USE_V2_BEHAVIOR:
         if self.fused is None:
-          self.fused = (ndims == 4)
-        elif self.fused and ndims != 4:
+          self.fused = (ndims == 4) or (ndims == 5)
```
[the remark this answers is not in the document]
Comment (kaixih) [reply tensorflow/python/keras/layers/normalization.py]: Done.

File: tensorflow/python/keras/layers/normalization_test.py (hunk 1 above)
Comment (kaixih) [reply tensorflow/python/keras/layers/normalization_test.py:331]: Done

File: tensorflow/python/ops/nn_grad.py (hunk 2 above)
Comment (kaixih) [reply tensorflow/python/ops/nn_grad.py]: Done.

File: tensorflow/core/grappler/optimizers/generic_layout_optimizer_transposer.cc [hunk 4]
```diff
@@ -654,7 +654,24 @@ Status LayoutSensitiveOpTransposer::UpdateNode(TransposeContext* context,
...
   DCHECK(IsDefaultLayoutSensitiveOp(*node->node()));
-  if (!ShouldProcess(*context, *node) || !IsFanoutPortRankN(*node, 0, 4)) {
+  const auto* output_shape_attr = node->GetAttr(kAttrOutputShape);
+  const auto& shape = output_shape_attr->list().shape(0);
+  const int rank = shape.dim_size();
+  std::string src_format = context->src_format;
+  std::string dst_format = context->dst_format;
+  // Update the format from 4D to 5D layout if necessary.
+  if (rank == 5) {
+    std::string src_format_3d = src_format == "NHWC" ? "NDHWC" : "NCDHW";
```
Comment (andyly) [inline tensorflow/core/grappler/optimizers/generic_layout_optimizer_transposer.cc:664]: Can this and the change below related to checking 4D data formats and upgrading to 5D data formats be updated to match https://github.com/tensorflow/tensorflow/pull/43226?

**Review by andyly — CHANGES_REQUESTED** [review]

File: tensorflow/core/grappler/optimizers/generic_layout_optimizer_transposer.cc (hunk 4 above)
Comment (kaixih) [reply tensorflow/core/grappler/optimizers/generic_layout_optimizer_transposer.cc:665]: Yes, they are related. Changed this PR by adding the same restriction on format checking.

**Review by andyly — DISMISSED** [review]

**Comment by rthadur** [comment]
@kaixih can you please resolve conflicts ?

**Comment by rmlarsen** [comment]
@reedwm @andyly Are your comments addressed to your satisfaction?

**Review by andyly — APPROVED** [review]

**Review by reedwm — DISMISSED** [review]

File: tensorflow/core/kernels/fused_batch_norm_op.cc [hunk 5]
```diff
@@ -1264,6 +1264,20 @@ class FusedBatchNormOpBase : public OpKernel {
...
+    TensorShape dest_shape;
+    if (use_reshape) {
+      const int64 in_batch = GetTensorDim(x, tensor_format_, 'N');
+      int64 in_planes = GetTensorDim(x, tensor_format_, '0');
+      int64 in_rows = GetTensorDim(x, tensor_format_, '1');
+      int64 in_cols = GetTensorDim(x, tensor_format_, '2');
+      const int64 in_depth = GetTensorDim(x, tensor_format_, 'C');
+      dest_shape = ShapeFromFormat(
+          tensor_format_, in_batch, {{in_planes, in_rows * in_cols}}, in_depth);
+      CHECK(x.CopyFrom(x, dest_shape));
```
Comment (andyly) [inline tensorflow/core/kernels/fused_batch_norm_op.cc:1278]: Can these `CHECK`s be converted to `OP_REQUIRES`? Currently that is triggering one of our presubmits recommending against using CHECK.

**Review by andyly — CHANGES_REQUESTED** [review]

File: tensorflow/core/kernels/fused_batch_norm_op.cc (hunk 5 above)
Comment (kaixih) [reply tensorflow/core/kernels/fused_batch_norm_op.cc]: Done. PTAL.

**Review by andyly — DISMISSED** [review]

**Comment by gbaned** [comment]
@kaixih Can you please resolve conflicts? Thanks!

**Comment by kaixih** [comment]
The conflicts are resolved. PTAL.

**Review by andyly — APPROVED** [review]
